staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -909,6 +910,30 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 		}
 	}
 
+	if len(schema.XValidations) > 0 {
```

Comment:
along the same lines, https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax indicates minimal bounds for implementations:

> Implementations are required to support at least:
>
> * 32 levels of nested expressions;
> * 32 repetitions of self-recursive or repetitive rules, i.e.:
> * 32 terms separated by || in a row;
> * 32 terms separated by && in a row;
> * 32 relations in a row;
> * 32 binary arithmetic operators of the same precedence in a row;
> * 32 selection (.) operators in a row;
> * 32 indexing ([_]) operators in a row;
> * 32 function call arguments;
> * list literals with 32 elements;
> * map or message literals with 32 fields.

does cel-go limit itself to those minimums? are we using any of the [heuristic cost estimations](https://github.com/google/cel-go/blob/master/interpreter/coster.go#L19-L22) to bound complexity of the expressions we accept?

Reply:
it might be easier to do validation that doesn't require cel expression parsing here...

* make sure `len(strings.TrimSpace(rule)) > 0` on all items
* make sure `len(strings.TrimSpace(message)) > 0` if `len(message) > 0` (can't provide a white-space-only `message` value)
* maybe require `len(strings.TrimSpace(message)) > 0` if `strings.Contains(strings.TrimSpace(rule), "\n")` (see comment on multi-line rules not being particularly suitable for outputting in errors)

That would let you directly construct field paths to the exact fields that had non-cel-parsing problems instead of trying to plumb them back from cel.Compile

Reply:
Done. This did work out to be quite a bit simpler and cleaner. I still let validation continue into the compiler and skip any empty rules there.